### PR TITLE
refactor if so that the "then type" is an expression

### PR DESCRIPTION
**Pull request by nikomatsakis**
This is a useful refactoring that I want for other branches.

r? @eddyb

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -930,7 +930,7 @@ pub enum Expr_ {
     /// An `if` block, with an optional else block
     ///
     /// `if expr { block } else { expr }`
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: This comment should be updated.

**Review by eddyb — APPROVED** [review]
r=me with the comment and the title ("then type") fixed

**Comment by TimNN** [comment]
Travis failed: 
```
/checkout/src/librustc_typeck/check/mod.rs:2765: line longer than 100 chars
```

**Comment by nikomatsakis** [comment]
closing this PR; I'll roll the change into #40224

**Pull request closed without merging**